readme highlights and quick tour (#99)
vrtex-team/pytorch_geometric: README.md

* PyTorch Geometric is **easy-to-use and friendly for first-time users** of graph neural networks. All it takes is a few lines of code to get started with training a GNN model (see the next section on a [quick tour](#quick-tour-for-new-users).
* PyTorch Geometric is **fast-growing and actively maintained**. Most of the state-of-the-art graph neural network architectures have been implemented either by library developers or authors of research papers themselves. 
* The PyTorch Geometric  framework is **flexible** and can be easily extended for building your own GNNs as researchers.   Making modifications to existing models or creating a new architecture is very easy thanks to its easy-to-use message passing API, and a variety of utility functions.
* We focus on the need of GNN applications in challenging real-world scenarios, and support **creation of GNN models with a variety of capabilities**, including but not limited to: scalable GNNs that can train on billions of nodes; dynamic GNNs for graph prediction over time; heterogeneous GNNs for tackling multiple node types and edge types in the same graph. 

> 

Reviewer: This looks great! Have some thoughts based on my Readme experience. I guess the first thing for the bullet points is to keep them short with a callout:

PyTorch Geometric is featured for:

- **Easy-to-use and unified API**:  build graph neural networks for fewer than 10 lines of code!
- **Comprehensive and well maintained GNN models**: more than **x** state-of-the-art GNN models, along with rich documentation and tutorial
- **Great flexibility** for easy extension to new models and applications, for both researchers and practitioners
- other items...

It may also be helpful to call out some numbers, like fewer than "10 lines" and "x" GNN models. These numbers can be attractive :)

Just my 2 cents.

--------------------------------------------------------------------------------
### Quick Tour for New Users
In this quick tour, we highlight the ease of creating and training a GNN models with only a few lines of code.

#### Simple example of GNN training
In the first glimpse of PyG, we use a few lines of code to implement training of a GCN for node classification task on the standard benchmark Cora.

The example requires the following imports:
```python
import torch
import torch.nn.functional as F
from torch_geometric.datasets import Planetoid
from torch_geometric.nn import GCNConv
```

> We first load the Cora dataset (More information can be found for [benchmark datasets](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#common-benchmark-datasets) and [creating your own datasets](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html), and then create a simple 2-layer GCN model using the already implemented GCN layer (GCNConv).

Reviewer 2: I think we do not need the training loop here. We should probably just show how to build a GNN model and how data is inputted.

```python

> dataset  =  Planetoid(dataset='Cora', transform=T.NormalizeFeatures())

Reviewer 2: ```suggestion
dataset  =  Planetoid(name='Cora', transform=T.NormalizeFeatures())
```

# creating a GCN model using the GCN layer (GCNConv)
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = GCNConv(dataset.num_features, 16)
        self.conv2 = GCNConv(16, dataset.num_classes)

>     def forward(self):

Reviewer 2: ```suggestion
    def forward(self, x, edge_index):
```

        x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr

>         x = F.relu(self.conv1(x, edge_index, edge_weight))

Reviewer 2: ```suggestion
        x = self.conv1(x, edge_index, edge_weight).relu()
```

        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index, edge_weight)

>         return F.log_softmax(x, dim=1)

Reviewer 2: ```suggestion
        return x
```
and use `F.cross_entropy_loss` (feels a bit easier to read).

```
In the next step, we instantiate the model and optimize it in a training loop, similar to standard PyTorch training procedure:
```python
model = Net()
optimizer = torch.optim.Adam(model.parameters(), weight_decay=0, lr=0.01)

best_val_acc = test_acc = 0
for epoch in range(200):
    optimizer.zero_grad()
    loss = F.nll_loss(model()[data.train_mask], data.y[data.train_mask])
    loss.backward()
    optimizer.step()
    print('Epoch: {:03d}, Loss:{:.4f}'.format(epoch, loss))
```

> More information about evaluating the training, validation and test accuracy can be found in the [example](*https://github.com/vrtex-team/pytorch_geometric/blob/master/examples/gcn.py).

Reviewer: This is a great example for beginners!
I am thinking maybe we could do this:

- give the core definition of a GNN model 
- and point them to the file in the example folder

```
from torch_geometric.nn import GCNConv  # prebuilt GCN conv layer

class GCN(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = GCNConv(dataset.num_features, 16)
        self.conv2 = GCNConv(16, dataset.num_classes)
    def forward(self):
        x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
        x = F.relu(self.conv1(x, edge_index, edge_weight))
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index, edge_weight)
        return F.log_softmax(x, dim=1)

model = GCN()
```

So we could show people it is easy to define a GNN and we could safely skip `import torch` and training process as they are the standard operations :)

Author: makes sense! I hid them in the drop down, just for comprehensiveness.

--------------------------------------------------------------------------------

### Library Highlights
*PyTorch Geometric* is a library that allows users to easily write Graph Neural Networks (GNNs) for a wide range of applications related to structured data. 
Whether you are machine learning researcher or first-time user of machine learning toolkits, here are some reasons to try out PyTorch Geometric for machine learning on graph-structured data.

* **Easy-to-use and unified API**: All it takes is 10-20 lines of code to get started with training a GNN model (see the next section on a [quick tour](#quick-tour-for-new-users).

> * **Comprehensive and well-maintained GNN models**: Most of the state-of-the-art graph neural network architectures have been implemented either by library developers or authors of research papers themselves. 

Reviewer 3: ```suggestion
* **Comprehensive and well-maintained GNN models**: Most of the state-of-the-art graph neural network architectures have already been implemented by either library developers or authors of research papers and are ready to be applied. 
```

> * **Great flexibility**: Existing PyG models can be easily extended for building your own GNNs as researchers.   Making modifications to existing models or creating a new architecture is very easy thanks to its easy-to-use message passing API, and a variety of operators and utility functions.

Reviewer 3: ```suggestion
* **Great flexibility**: Existing PyG models can easily be extended for conducting your own research with GNNs. Making modifications to existing models or creating a new architecture is simple, thanks to its easy-to-use message passing API, and a variety of operators and utility functions.
```

* **Large-scale real-world GNN models**: We focus on the need of GNN applications in challenging real-world scenarios, and support learning on diverse types of graphs, including but not limited to: scalable GNNs for graphs with billions of nodes; dynamic GNNs for graph prediction over time; heterogeneous GNNs for tackling multiple node types and edge types in the same graph. 

--------------------------------------------------------------------------------
### Quick Tour for New Users
In this quick tour, we highlight the ease of creating and training a GNN models with only a few lines of code.

#### Simple example of GNN training

> In the first glimpse of PyG, we use a few lines of code to implement training of a GCN for node classification task on the standard benchmark Cora.

Reviewer 3: ```suggestion
In the first glimpse of PyG, we use a few lines of code to implement training of a GCN for a node classification task on the standard benchmark Cora.
```

<details>
<summary>We first import the necessary libraries</summary>

```python
import torch
import torch.nn.functional as F
from torch_geometric.datasets import Planetoid
from torch_geometric.nn import GCNConv
```
</details>

> We then load the Cora dataset (More information can be found for [benchmark datasets](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#common-benchmark-datasets) and [creating your own datasets](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html), and create a simple 2-layer GCN model using the already implemented GCN layer (GCNConv).

Reviewer 3: ```suggestion
We then load the Cora dataset (more information can be found for [benchmark datasets](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#common-benchmark-datasets) and [creating your own datasets](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html), and create a simple 2-layer GCN model using the already implemented GCN layer (GCNConv).
```

```python
dataset  =  Planetoid(name='Cora', transform=T.NormalizeFeatures())

# creating a GCN model using the GCN layer (GCNConv)
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = GCNConv(dataset.num_features, 16)
        self.conv2 = GCNConv(16, dataset.num_classes)

    def forward(self, x, edge_index):
        x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
        x = self.conv1(x, edge_index, edge_weight).relu()
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index, edge_weight)
        return x
```
<detail>

> <summary>A graph neural network model has been created! We can then instantiate the model and optimize it in a training loop, similar to standard PyTorch training procedure.</summary>

Reviewer 3: ```suggestion
<summary>A graph neural network model has been created! We can now instantiate the model and optimize it in a training loop, similar to the standard PyTorch training procedure.</summary>
```

```python
model = Net()
optimizer = torch.optim.Adam(model.parameters(), weight_decay=0, lr=0.01)

best_val_acc = test_acc = 0
for epoch in range(200):
    optimizer.zero_grad()
    loss = F.nll_loss(model()[data.train_mask], data.y[data.train_mask])
    loss.backward()
    optimizer.step()
    print('Epoch: {:03d}, Loss:{:.4f}'.format(epoch, loss))
```
</detail>
More information about evaluating the training, validation and test accuracy can be found in the [example](*https://github.com/vrtex-team/pytorch_geometric/blob/master/examples/gcn.py).

#### Creating your own GNNs

> In addition to using an existing GNN, PyTorch Geometric also makes implementing Graph Neural Networks a breeze (see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html) for the accompanying tutorial).

Reviewer 3: ```suggestion
In addition to the easy application of existing GNNs, PyTorch Geometric makes it simple to implement custom Graph Neural Networks (see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html) for the accompanying tutorial).
```

        edge_features = torch.cat([x_i, x_j - x_i], dim=1)  # shape [E, 2 * F_in]
        return self.mlp(edge_features)  # shape [E, F_out]
```

> Users are highly encouraged to check out the [documentation](https://pytorch-geometric.readthedocs.io/en/latest/), which contains a longer 60-minute tutorial on the essential functionalities of PyTorch Geometric. In addition, users can find in documentations the advanced functionalities, such as heterogeneous GNNs, scalable GNNs, minibatch sampling etc.

Reviewer 3: ```suggestion
Users are highly encouraged to check out the [documentation](https://pytorch-geometric.readthedocs.io/en/latest/), which contains a longer 60-minute tutorial on the essential functionalities of PyTorch Geometric, including advanced features, such as GNNs on heterogeneous graphs, scalable GNNs, and minibatch sampling.
```

--------------------------------------------------------------------------------

### PyG Architecture Overview

PyTorch Geometric provides a multi-layer framework that enables users to build graph neural network solutions on both low and high levels. It comprises of the following components: Engine, Storage, Operators and GNN Models.

* **PyG Engine** is built based on the powerful PyTorch deep learning framework, as well as additions of efficient CUDA libraries that are commonly used in GNNs, such as torch_scatter and torch_sparse.

> * **PyG Storage** handles data processing, loading and transformation pipelines. It is capable of handling and processing large-scale graph datasets, and provide effective solutions for dynamic and heterogeneous graphs. To further enhance scalability, it provides a variety of sampling methods, which enable training of GNNs on huge web-scale graphs.

Reviewer 3: ```suggestion
* **PyG Storage** handles data processing, loading and transformation pipelines. It is capable of handling and processing large-scale graph datasets, and provides effective solutions for dynamic and heterogeneous graphs. To further enhance scalability, it provides a variety of sampling methods, which enable training of GNNs on huge web-scale graphs.
```

* **PyG Operators** is a collection of essential functionalities for a variety of graph neural networks. These are important components that can be combined and applied to various parts of a GNN model, ensuring the flexibility of GNN design.

> * Finally, PyG provides abundant set of **GNN models**, as well as examples that showcases the models on standard graph benchmarks. Thanks to its flexibility, users can easily build and modify custom GNN models to fit their specific needs when needed.

Reviewer 3: ```suggestion
* Finally, PyG provides an abundant set of **GNN models** and examples that showcases the models on standard graph benchmarks. Thanks to its flexibility, users can easily build and modify custom GNN models to fit their specific needs.
```

![Architecture](https://github.com/vrtex-team/pytorch_geometric/blob/readme/architecture.jpg)

--------------------------------------------------------------------------------
### Implemented GNN models
In this section, we list a variety of GNN models, layers and operators according to categories.

<details>
<summary><b>GNN Layers</b>

> The graph neural network layers are implemented by inheriting the **nn.MessagePassing** class. Each type of GNN layers specify different ways to perform message passing, e.g. by designing different message function, aggregation function and update functions as defined in the [documentation](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html#). These GNN layers can be stacked together to create graph neural network models.

Reviewer 3: ```suggestion
The graph neural network layers are implemented by inheriting from the **nn.MessagePassing** class. A GNN layer specifies how to perform message passing, e.g. by designing different message functions, aggregation functions and update functions as defined in the [documentation](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html#). These GNN layers can be stacked together to create graph neural network models.
```

* **[PANPooling](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.pool.PANPooling)** from Ma *et al.*: [Path Integral Based Convolution and Pooling for Graph Neural Networks](https://arxiv.org/abs/2006.16811) (NeurIPS 2020)
* **[MemPooling](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.pool.MemPooling)** from Khasahmadi *et al.*: [Memory-Based Graph Networks](https://arxiv.org/abs/2002.09518) (ICLR 2020) [[**Example**](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/mem_pool.py)]
* **[Local Degree Profile](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.transforms.LocalDegreeProfile)** from Cai and Wang: [A Simple yet Effective Baseline for Non-attribute Graph Classification](https://arxiv.org/abs/1811.03508) (CoRR 2018)
</details>

<details>
<summary><b>Models</b>

> Here a collection of graph learning models are implemented. Unlike GNN layers, these models have already been built with multiple message passing layers, and users can directly use the model to make predictions on graphs. Unlike simple stacking of GNN layers, these models could involve preprocessing, additional learnable parameters, skip layers, graph coarsening etc. These pre-built GNN models makes applying state-of-the-art models a breeze.

Reviewer 3: ```suggestion
In the following, implemented graph learning models are listed. Unlike GNN layers, these models have already been built with multiple message passing layers, and users can directly use the model to make predictions on graphs. Unlike simple stacking of GNN layers, these models could involve preprocessing, additional learnable parameters, skip layers, graph coarsening etc. These pre-built GNN models make applying state-of-the-art models a breeze.
```

* **[GNNExplainer](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.models.GNNExplainer)** from Ying *et al.*: [GNNExplainer: Generating Explanations for Graph Neural Networks](https://arxiv.org/abs/1903.03894) (NeurIPS 2019) [[**Example**](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/gnn_explainer.py)]
* **[Node2Vec](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.models.Node2Vec)** from Grover and Leskovec: [node2vec: Scalable Feature Learning for Networks](https://arxiv.org/abs/1607.00653) (KDD 2016) [[**Example**](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/node2vec.py)]
</details>

<details>
<summary><b>GNN Operators and Utilities</b>

> PyG comes with a rich set of neural network operators that are commonly used in many GNN models. PyG is built with extensible design: it is easy to apply these operators and graph utilities to the existing GNN layers and models, to further enhance model performance.

Reviewer 3: ```suggestion
PyG comes with a rich set of neural network operators that are commonly used in many GNN models. They follow an extensible design: it is easy to apply operators and graph utilities to the existing GNN layers and models, to further enhance model performance.
```

* **[CorrectAndSmooth](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.models.CorrectAndSmooth)** from Huang *et al.*: [Combining Label Propagation And Simple Models Out-performs Graph Neural Networks](https://arxiv.org/abs/2010.13993) (CoRR 2020) [[**Example**](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/correct_and_smooth.py)]
* **[Gini](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.functional.gini)** and **[BRO](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.functional.bro)** regularization from Henderson *et al.*: [Improving Molecular Graph Neural Network Explainability with Orthonormalization and Induced Sparsity](https://arxiv.org/abs/2105.04854) (ICML 2021)
</details>

<details>
<summary><b>Scalable GNNs</b>

> PyG supports implementation of graph neural networks that can scale to extremely large graphs (e.g. billions of nodes and edges). Such application is challenging because the entire graph, its associated features and the GNN parameters cannot fit into GPU memory. Many state-of-the-art scalable GNNs tackle the challenge by (1) sampling neighborhood for minibatch training; (2) graph clustering and partitioning (3) using simplied GNN models. These approaches have been implemented, and can benefit from the above GNN layers, operators and models.

Reviewer 3: ```suggestion
PyG supports implementation of graph neural networks that can scale to extremely large graphs (e.g. billions of nodes and edges). Such application is challenging because the entire graph, its associated features and the GNN parameters cannot fit into GPU memory. Many state-of-the-art scalable GNNs tackle the challenge by (1) sampling neighborhood for minibatch training, (2) graph clustering and partitioning, or (3) using simplified GNN models. These approaches have been implemented, and can benefit from the above GNN layers, operators and models.
```
